### PR TITLE
Enable link-time optimisations on Release build.

### PR DESCRIPTION
**Pull request by Yamakaky**
Fixes https://github.com/neovim/neovim/issues/3136

### CODE REVIEW

**Comment by Yamakaky** [comment]
It seems that the ci only tests Debug ?

**Comment by justinmk** [comment]
> It seems that the ci only tests Debug ?

Yeah I was afraid of that. Long ago we tried to add it but ran into some problems. Since we are shipping Release now I don't think those problems exist, so we should also add a travis build for it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -82,6 +82,12 @@ if(CMAKE_C_FLAGS_RELEASE MATCHES "-O3")
   string(REPLACE "-O3" "-O2" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
 endif()
 
+# Enable link-time optimisations on release builds.
+check_c_compiler_flag(-flto HAS_FLTO_FLAG)
+if(HAS_FLTO_FLAG)
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS} -flto")
```
Comment (justinmk) [inline CMakeLists.txt]: We should also enable it for RelWithDebInfo.

Based on [this commit](https://github.com/poelzi/ulatencyd/commit/32b7684ea919892ad63e1f1449acda7a2fed95b9#diff-af3b638bc2a3e6c650974192a53c7291R171) and other blurbs around the web, LTO needs the compiler flags to be _the same_ as the the linker flags.

CMAKE_EXE_LINKER_FLAGS_RELEASE
CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO

If that's true then I wonder why it seems to have worked for @mhinz?

Comment (mhinz) [inline CMakeLists.txt]: Maybe it's only needed for gcc? I (and probably most OSX users?) usually build using clang.

For my quick test I simply added `-lfto` to this line: https://github.com/neovim/neovim/blob/master/CMakeLists.txt#L144

Comment (mhinz) [inline CMakeLists.txt]: Confirmed, only adding to `CMAKE_C_FLAGS` leads to errors like this in the Release build:

```
[103/108] Building C object src/nvim/CMakeFiles/nvim.dir/regexp.c.o
clang: warning: -lfto: 'linker' input unused
```

**Comment by mhinz** [comment]
Hmm, the current patch won't break OSX build, but it will skip the `-lfto` flag. I look into it.

**Comment by mhinz** [comment]
Sorry for the confusion above. I still had another flag in the CMakeLists.txt.

NOTE: `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc` is clang.

### clang, without LTO

```
$ nvim --version
NVIM v0.1.3-151-gfc51f86
Build type: Release
Compilation: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/data/repo/neovim/build/config -I/data/repo/neovim/src -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include/luajit-2.0 -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/usr/local/opt/gettext/include -I/usr/include -I/data/repo/neovim/build/src/nvim/auto -I/data/repo/neovim/build/include
[...]

$ du -b build/bin/nvim
3076020 build/bin/nvim
```

### clang, with LTO

```
$ git diff -U0
diff --git i/CMakeLists.txt w/CMakeLists.txt
index 527a085..4d18067 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -84,0 +85,5 @@ endif()
+check_c_compiler_flag(-flto HAS_FLTO_FLAG)
+if(HAS_FLTO_FLAG)
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -flto")
+endif()

$ nvim --version
NVIM v0.1.3-151-gfc51f86
Build type: Release
Compilation: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -flto -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/data/repo/neovim/build/config -I/data/repo/neovim/src -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include/luajit-2.0 -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/data/repo/neovim/.deps/usr/include -I/usr/local/opt/gettext/include -I/usr/include -I/data/repo/neovim/build/src/nvim/auto -I/data/repo/neovim/build/include

$ du -b build/bin/nvim
3430500 build/bin/nvim
```

So, with LTO the binary got half a megabyte bigger.

**Comment by justinmk** [comment]
Thanks for investigating. I guess we need to make sure the linker is using _all_ of the same flags as the compiler. Do we know if that's the case currently?

**Comment by mhinz** [comment]
So I read [this](http://stackoverflow.com/questions/24718652/why-does-link-time-optimization-results-in-larger-binaries). The `-Os` seems to be needed.

| Flags for Release build | Size in Bytes |
|--------------------------------|-------------------|
| `<no addition>` | `3076020` |
| `-flto` | `3430500` |
| `-Os` | `2726328` |
| `-flto -Os` | `2682564` |

**Comment by justinmk** [comment]
Nice! If I understand correctly, `-Os` is needed in that case _because_ it was used in the compiler flags. I could be wrong but I thought MinSizeRel used `-Os` which precluded `-O2`. 

So for each case, we just need to make sure the same flags are used by the compiler and linker? If `-Os` is _always_ required for `flto` then we should only use it for MinSizeRel... right?

**Comment by mhinz** [comment]
According to https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html it should be the same _optimization_ flags for compile and link time. So `-O2` for Release (although it's increasing binary size) and `-Os` for MinSizeRel.

But maybe we should go for this instead: https://cmake.org/cmake/help/v3.0/prop_tgt/INTERPROCEDURAL_OPTIMIZATION.html

Interprocedural optimization is a set of techniques, including LTO: https://en.wikipedia.org/wiki/Interprocedural_optimization

**Comment by Yamakaky** [comment]
`-flto` isn't only used for binary size, it enables optimisations like cross .o inlining for example.

**Comment by justinmk** [comment]
Great, seems like we understand the details now. We should probably revert this change until we get everything correct.

**Comment by Yamakaky** [comment]
https://cmake.org/cmake/help/v3.0/prop_tgt/INTERPROCEDURAL_OPTIMIZATION_CONFIG.html

**Comment by mhinz** [comment]
Yea, it dumps some kind of AST to disk for all translation units, so it can apply optimizations as if there was only one big translation unit afterwards.

So, I checked `cmake -LAH` in the Neovim repo. Neither cmake itself nor our CMakeLists.txt sets any _optimization_ option for link time. Thus this should be the proper approach:

```cmake
check_c_compiler_flag(-flto HAS_FLTO_FLAG)
if(HAS_FLTO_FLAG)
  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -flto")
  set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -flto")
  set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} -flto")
  set(CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} -flto")
endif()
```

But our CMakeLists.txt also does this:

```cmake
# Default to -O2 on release builds.
if(CMAKE_C_FLAGS_RELEASE MATCHES "-O3")
  message(STATUS "Replacing -O3 in CMAKE_C_FLAGS_RELEASE with -O2.")
  string(REPLACE "-O3" "-O2" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
endif()
```

So the proper way would be:

```cmake
check_c_compiler_flag(-flto HAS_FLTO_FLAG)
if(HAS_FLTO_FLAG)
  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -flto")
  set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -flto")
  set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} -flto -O2")
  set(CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} -flto")
endif()
```

(Mind the additional `-O2` in CMAKE_EXE_LINKER_FLAGS_RELEASE.)

**Comment by Yamakaky** [comment]
I can't make INTERPROCEDURAL_OPTIMIZATION to work... With `make SHELL=/bin/bash -x`, I can see that the command line doesn't change.

**Comment by mhinz** [comment]
According to the discussion I posted above, it's only implemented for the Intel compiler at the moment.

**Comment by ZyX-I** [comment]
How about using INTERPROCEDURAL_OPTIMIZATION and running suggestion from https://github.com/neovim/neovim/pull/4313#issuecomment-186922575 if this option was enabled, but compiler is not intel.

**Pull request merged**